### PR TITLE
Start a new TPU interpret mode for Pallas

### PR DESCRIPTION
**Pull request by jburnim**
The goal of this interpret mode is to run a Pallas TPU kernel on CPU, while simulating a TPU's shared memory, multiple devices/cores, remote DMAs, and synchronization.

The basic approach is to execute the kernel's Jaxpr on CPU, but to replace all load/store, DMA, and synchronization primitives with io_callbacks to a Python functions that simulate these primitives. When this interpret mode is run inside of shard_map and jit, the shards will run in parallel, simulating the parallel execution of the kernel on multiple TPU devices.

The initial version in this PR can successfully interpret the examples in https://jax.readthedocs.io/en/latest/pallas/tpu/distributed.html , but is still missing a lot of functionality, including:

 - Executing DMAs asynchronously.

 - Padding in pallas_call.

 - Propagating source info.

### CODE REVIEW

File: tests/pallas/tpu_pallas_interpret_test.py [hunk 1]
```diff
@@ -0,0 +1,974 @@
...
+            [pltpu.SemaphoreType.DMA] * 2
+        ),
+    )
+    right_permute = pl.pallas_call(
+        right_permute_kernel,
+        out_shape=out_shape,
+        grid_spec=grid_spec,
+        compiler_params=pltpu.TPUCompilerParams(collective_id=13),
+        interpret=True,
+        backend='mosaic_tpu',
```
Comment (jburnim) [inline tests/pallas/tpu_pallas_interpret_test.py:113]: Idea: backend="cpu_like_mosaic_tpu"

File: jax/_src/pallas/mosaic/interpret.py [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+
+  next_buffer_id: dict = dataclasses.field(
+    default_factory=lambda: collections.defaultdict(lambda: 100))
+  next_semaphore_id: dict = dataclasses.field(
+    default_factory=lambda: collections.defaultdict(lambda: 2000))
+
+shared_memory = SharedMemory()
+
+class Semaphore:
+  def __init__(self, num_devices=1):
```
Comment (justinjfu) [inline jax/_src/pallas/mosaic/interpret.py:78]: Where is num_devices used?

File: jax/_src/pallas/mosaic/interpret.py [hunk 3]
```diff
@@ -0,0 +1,694 @@
...
+
+  # TODO(jburnim): Check/fix so that IDs for barrier semaphores do not conflict
+  # with IDs for regular or DMA semaphores.  (For example, store them in a
+  # different table.)
+  with shared_memory.lock:
+    semaphore_id = collective_id
+    if not semaphore_id in shared_memory.sem:
+      shared_memory.sem[semaphore_id] = Semaphore()
+
+  return np.int16(semaphore_id)
```
Comment (justinjfu) [inline jax/_src/pallas/mosaic/interpret.py:144]: Why does this need to be int16?

File: jax/_src/pallas/pallas_call.py [hunk 4]
```diff
@@ -1878,8 +1878,10 @@ def wrapped(*args):
             f"a different abstract value {out_aval}.")
 
     index_args, rest_args = split_list(flat_args, [grid_mapping.num_index_operands])
-    with pallas_core.interpret_mode_env(interpret):
-      out_flat = pallas_call_p.bind(
+    if interpret and backend == 'mosaic_tpu':
```
Comment (justinjfu) [inline jax/_src/pallas/pallas_call.py:1881]: Can we just add this as an alternative backend to pallas_call and register it in the MLIR lowering? I think it's confusing to have multiple interpret modes (interpret itself isn't even a great name). You would probably just have to call MLIR's lower_fun on your interpreter.

File: jax/_src/pallas/mosaic/interpret.py [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+    invals = jax.util.safe_map(read, eqn.invars)
+
+    if prim is primitives.load_p:
+      raise NotImplementedError()
+
+    elif prim is lax.cond_p:
+      def _make_branch(jaxpr):
+        return lambda *args: _interpret_jaxpr(
+            jaxpr, *args, compiler_params=compiler_params)
+      out = lax.switch(
```
Comment (justinjfu) [inline jax/_src/pallas/mosaic/interpret.py:343]: Why call switch instead of calling cond again?

File: jax/_src/pallas/mosaic/interpret.py [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+  # TODO(jburnim): Do this transform with NumPy arrays?
+  result, result_val = state_discharge.transform_swap_array(
+      jnp.array(mem_val), transforms, jnp.array(val))
+  result_val = np.array(result_val)
+  with shared_memory.lock:
+    shared_memory.mem[(memory_space, buffer_id, device_id)] = result_val
+
+  return np.array(result)
+
+def execute_dma(src, dst, send_sem, recv_sem):
```
Comment (justinjfu) [inline jax/_src/pallas/mosaic/interpret.py:212]: Add type annotations for src/dst?

File: jax/_src/pallas/mosaic/interpret.py [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+BlockSpecTree = pallas_core.BlockSpecTree
+NoBlockSpec = pallas_core.NoBlockSpec
+no_block_spec = pallas_core.no_block_spec
+ScratchShapeTree = pallas_core.ScratchShapeTree
+CostEstimate = pallas_core.CostEstimate
+
+
+@dataclasses.dataclass(frozen=True)
+class SharedMemory:
+  # (memory_space, buffer_id, device_id) -> NumPy array
```
Comment (justinjfu) [inline jax/_src/pallas/mosaic/interpret.py:62]: This might not generalize to megacore, since I believe the HBM is shared but VMEM is not between cores. Not sure how you would split that up here. But probably fine to leave as a TODO.

File: tests/pallas/tpu_pallas_interpret_test.py (hunk 1 above)
Comment (justinjfu) [reply tests/pallas/tpu_pallas_interpret_test.py:113]: Yea that seems good. Or "tpu_verification" if the main purpose will be for verification.

**Review by justinjfu — APPROVED** [review]

File: jax/_src/pallas/mosaic/interpret.py (hunk 2 above)
Comment (jburnim) [reply jax/_src/pallas/mosaic/interpret.py:78]: Good catch -- removed.

(I think we will want this argument if we change the representation of Semaphore to an array of counts, instead of a dictionary.  But that will be a later change.)

File: jax/_src/pallas/mosaic/interpret.py (hunk 3 above)
Comment (jburnim) [reply jax/_src/pallas/mosaic/interpret.py:144]: My thinking was that it might be useful to match SEMAPHORE_INTERPRET_DTYPE in `pallas/core.py`:
```python
# Datatype for semaphore values in interpret mode.
# For now, we choose a relatively uncommon datatype (i16) so it is more easily
# identifiable in kernels.
# TODO(justinfu): Handle semaphores with a custom extended dtype.
SEMAPHORE_INTERPRET_DTYPE = jnp.int16
```

I'll add a comment noting this.

(I'm also happy to switch this dtype if you think int32 or something else makes more sense.)

File: jax/_src/pallas/mosaic/interpret.py (hunk 5 above)
Comment (jburnim) [reply jax/_src/pallas/mosaic/interpret.py:343]: It looks to me like `cond_p` is actually a switch -- i.e., `invals[0]` is an int that indexes the elements of`eqn.params['branches']`, and there could be more than 2 elements.

File: jax/_src/pallas/mosaic/interpret.py (hunk 6 above)
Comment (jburnim) [reply jax/_src/pallas/mosaic/interpret.py:212]: I added a comment for now.

I think this function may go away when the code is updated to execute DMAs asynchronously.

File: jax/_src/pallas/mosaic/interpret.py [hunk 8]
```diff
@@ -0,0 +1,906 @@
...
+  # TODO(jburnim): Handle Megacore.
+  mem: dict = dataclasses.field(default_factory=dict)
+
+  # semaphore_id -> Semaphore
+  sem: dict = dataclasses.field(default_factory=dict)
+
+  lock: threading.Lock = dataclasses.field(default_factory=threading.Lock)
+
+  next_buffer_id: dict = dataclasses.field(
```
Comment (sharadmv) [inline jax/_src/pallas/mosaic/interpret.py:99]: do we risk these ids clobbering each other at some point?

File: jax/_src/pallas/mosaic/interpret.py [hunk 9]
```diff
@@ -0,0 +1,906 @@
...
+
+  shared_memory = _get_shared_memory()
+  with shared_memory.lock:
+    semaphore_id = shared_memory.next_semaphore_id[device_id]
+    shared_memory.next_semaphore_id[device_id] = semaphore_id + num_semaphores
+    for i in range(semaphore_id, semaphore_id + num_semaphores):
+      if not i in shared_memory.sem:
+        shared_memory.sem[i] = Semaphore()
+
+  # NOTE: For now, we use a relatively uncommon datatypen (int16) for
```
Comment (sharadmv) [inline jax/_src/pallas/mosaic/interpret.py:158]: nit: datatypen -> datatype

File: jax/_src/pallas/mosaic/interpret.py [hunk 10]
```diff
@@ -0,0 +1,906 @@
...
+    prim = eqn.primitive
+    invals = jax.util.safe_map(read, eqn.invars)
+
+    if prim is primitives.load_p:
+      raise NotImplementedError()
+
+    elif prim is primitives.swap_p:
+      raise NotImplementedError()
+
+    elif prim is lax.cond_p:
```
Comment (sharadmv) [inline jax/_src/pallas/mosaic/interpret.py:436]: Quick comment about code organization. Conventionally we use a rules dict instead of a big if-elif chain. No need to modify this PR to do this but that would be nice to change later.

File: jax/_src/pallas/pallas_call.py [hunk 11]
```diff
@@ -82,19 +84,31 @@ def _jit_run(*args):
...
+    **params
 ):
   del avals
+
+  if isinstance(interpret, mosaic_tpu_interpret.TPUInterpretParams):
+    # Report effects that will be introduced when running/lowering
+    # mosaic_tpu_interpret.mosaic_tpu_interpret.interpret_pallas_call .
+    effs = mosaic_tpu_interpret.get_interpret_effects()
+  else:
+    effs = []
```
Comment (sharadmv) [inline jax/_src/pallas/pallas_call.py:100]: effects should be a set, not a list (can use core.no_effects IIRC for no effects).

File: tests/pallas/tpu_pallas_interpret_distributed_test.py [hunk 12]
```diff
@@ -0,0 +1,995 @@
...
+
+P = jax.sharding.PartitionSpec
+
+
+class InterpretDistributedTest(jtu.JaxTestCase):
+
+  def test_right_permute_example(self):
+    num_devices = jax.device_count()
+    if num_devices < 4:
+      self.skipTest(f'requires at least 4 devices, found {num_devices}')
```
Comment (sharadmv) [inline tests/pallas/tpu_pallas_interpret_distributed_test.py:44]: could move this boilerplate into setUp() and avoid repeating it in every test.

**Review by sharadmv — APPROVED** [review]
Is it possible to add a test that detects when there is a race?

File: jax/_src/pallas/mosaic/interpret.py (hunk 8 above)
Comment (jburnim) [reply jax/_src/pallas/mosaic/interpret.py:99]: We're using int16's for buffer IDs, so if one kernel allocates more than 65,536+ buffers per device, then we'll end up re-using IDs.  This might not be that hard to do with run_scoped.  A similar issues exists for semaphores.  (I'll add a TODO to raise an error if any of these IDs get too large to for int16.  And we could switch to uint32 to make this much less likely.)

But note that semaphore IDs and buffer IDs are used in different places, so it's fine if we have a buffer and a semaphore that end up with the same numeric ID.  (I just started IDs at 100 and 2000 so that, for small kernels, you can easily tell if an ID is for a buffer or a semaphore based on its size.)

(There is a potential issue with conflicting IDs for barrier semaphores vs. regular and DMA semaphores.  There is a TODO to prevent this.)

File: jax/_src/pallas/mosaic/interpret.py (hunk 9 above)
Comment (jburnim) [reply jax/_src/pallas/mosaic/interpret.py]: Good catch -- fixed.

File: jax/_src/pallas/mosaic/interpret.py (hunk 10 above)
Comment (jburnim) [reply jax/_src/pallas/mosaic/interpret.py:439]: 👍   There is a TODO to replace this with a dict of rules.

File: jax/_src/pallas/pallas_call.py (hunk 11 above)
Comment (jburnim) [reply jax/_src/pallas/pallas_call.py]: Fixed.

**Comment by jburnim** [comment]
> Is it possible to add a test that detects when there is a race?

I'll add some tests like this when I add the race detector in a later change.
